### PR TITLE
PLT-7932 Add tracking of the chain tip

### PR DESCRIPTION
**Pull request by berewt**
- The processing of the chain-sync events is modified to issue some events with the chain tip.
- We introduce the `LastEventIndexer`, an indexer that saves the last value of an event on disk, to use it to save chain tips.
- We modify the indexers organisation accordingly.

Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [ ] Commit sequence broadly makes sense and have useful messages
    - [ ] Important changes are reflected in changelog.d of the affected packages
    - [ ] Relevant tickets are mentioned in commit messages
- PR
    - [ ] (For external contributions) Corresponding issue exists and is linked in the description
    - [ ] Targeting main unless this is a cherry-pick backport
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] If relevant, reference the ADR in the PR and reference the PR in the ADR
    - [ ] Reviewer requested

### CODE REVIEW

File: marconi-chain-index/src/Marconi/ChainIndex/Experimental/Indexers/ChainTip.hs [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+import Marconi.ChainIndex.Experimental.Indexers.Orphans ()
+import Marconi.Core qualified as Core
+
+type instance Core.Point C.ChainTip = C.ChainPoint
+
+type ChainTipIndexer m = Core.WithTrace m Core.LastEventIndexer C.ChainTip
+
+data ChainTipConfig = ChainTipConfig
+  { path :: FilePath
+  -- ^ path to save infortmation about the chain tip
```
Comment (koslambrou) [inline marconi-chain-index/src/Marconi/ChainIndex/Experimental/Indexers/ChainTip.hs:42]: infortmation -> information

File: marconi-core/src/Marconi/Core/Indexer/LastEventIndexer.hs [hunk 2]
```diff
@@ -0,0 +1,266 @@
+{-# LANGUAGE FlexibleContexts #-}
```
Comment (koslambrou) [inline marconi-core/src/Marconi/Core/Indexer/LastEventIndexer.hs:1]: This is basically a FileIndexer, but for the last event only, right? We couldn't extend FileIndexer?

File: marconi-chain-index/src/Marconi/ChainIndex/Experimental/Indexers/ChainTip.hs [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+{-# LANGUAGE OverloadedStrings #-}
+
+{- |
+   An indexer that tracks the chainTip of the ledger, storing the tip in a file.ch
+
+   This indexer totally ignores the point it receives and uses the chain tip as a point in its file indexing.
+
+   Indeed, the notion of @lastSyncPoint@ and last stable point isn't relevant for this indexer as
+   the information it stores is updated at each block.
+-}
```
[the remark this answers is not in the document]
Comment (berewt) [reply marconi-chain-index/src/Marconi/ChainIndex/Experimental/Indexers/ChainTip.hs:12]: Yes, my bad, I decided to create the `LastEventIndexer` latein the PR and forgot to update the comments.

File: marconi-chain-index/src/Marconi/ChainIndex/Experimental/Indexers/ChainTip.hs [hunk 4]
```diff
@@ -0,0 +1,131 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+{- |
+   An indexer that tracks the chainTip of the ledger, storing the tip in a file.ch
```
[the remark this answers is not in the document]
Comment (berewt) [reply marconi-chain-index/src/Marconi/ChainIndex/Experimental/Indexers/ChainTip.hs:6]: It's fat finger mostly, there wasn't supposed to be any extension in the comment. :)

File: marconi-chain-index/src/Marconi/ChainIndex/Experimental/Runner.hs [hunk 5]
```diff
@@ -135,7 +138,13 @@ mkEventStream q =
 
       processEvent
         :: ChainSyncEvent BlockEvent
-        -> Core.ProcessedInput C.ChainPoint (WithDistance BlockEvent)
-      processEvent (RollForward x ct) = Core.Index $ Just <$> blockTimed x ct
-      processEvent (RollBackward x _) = Core.Rollback x
-   in S.mapM_ $ STM.atomically . STM.writeTBQueue q . processEvent
+        -> [Core.ProcessedInput C.ChainPoint (Either C.ChainTip (WithDistance BlockEvent))]
```
[the remark this answers is not in the document]
Comment (berewt) [reply marconi-chain-index/src/Marconi/ChainIndex/Experimental/Runner.hs:187]: Yes, I was hesitating between this and a sum type, I'd be happy to change it to sum type if you think it's clearer.

File: marconi-core/src/Marconi/Core/Indexer/LastEventIndexer.hs (hunk 2 above)
Comment (berewt) [reply marconi-core/src/Marconi/Core/Indexer/LastEventIndexer.hs:1]: We could, and this is how it was done initially, but having it in its own module streamlines a lot the logic and the way we process the events. If we want to combine both (and we probably want to chat about it first), I would prefer it to be a separate ticket.

Comment (koslambrou) [reply marconi-core/src/Marconi/Core/Indexer/LastEventIndexer.hs:1]: No problem. I'll create a ticket

**Comment by berewt** [comment]
@koslambrou I've asked for another review to be sure that the migration from `Either` to `TipOrBlockEvent` is ok for you.

**Review by koslambrou — APPROVED** [review]
LGTM :)

**Pull request merged**